### PR TITLE
runtime env prototype

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?
A runtime env pluggability and API prototype

### Motivation
Currently, we already have a [runtime env plugin](https://github.com/ray-project/ray/blob/master/python/ray/_private/runtime_env/plugin.py)  and some runtime env modules e.g. pip, conda, working_dir, container and java_jars. But the existing modules are all build-in which doesn't use the pluggability.

In long term, we aim to covert all the runtime env modules to plugins. So, we make this prototype to make the pluggability and API clear.

### Direction
- This prototype contains the entire plugin API and runtime env API. 
- We modify the plugin API to make it more reasonable and common.
- Every plugin should define its [json schema](https://json-schema.org/). 
- We define a RuntimeEnv API class in different languages. This class is a sdk for users to construct a runtime environment.
- We define a way to extend  RuntimeEnv in different language workers.
- The main work flow is in the [test_runtime_env.py](https://github.com/ray-project/ray/pull/25387/files#diff-a37369be55591323e18e53fb7d15abb3f970fe66b75998835ef410ac929f7d93)

A common work steps:
- First, we have a plugin interface [plugin.py](https://github.com/ray-project/ray/pull/25387/files#diff-a331b5855f9bdd43ca140bb15a371b4ed866ea212f66bf494a1d2ca123d26a28), a plugin manager [plugin_manager.py](https://github.com/ray-project/ray/pull/25387/files#diff-5862b6ec93c00728bf65cb7761c2dee9496d76717a0eda30889e18ef9d5b67cf), plugin schema manager in different languages(python [plugin_schema_manager.py](https://github.com/ray-project/ray/pull/25387/files#diff-5edab082e82f6111ab86e8a4b3586f28d06211a21065a752c008b4ae2b4fa2ea), java [PluginSchemaManager.java](https://github.com/ray-project/ray/pull/25387/files#diff-c56f45788f7613bdd36bbfb95fbaa114c2ec199ec5986cbaa4a8e077e2e62fc8), C++ [plugin_schema_manager.h](https://github.com/ray-project/ray/pull/25387/files#diff-153cac687ac2e4bae91656895fcadc3717e1aa3129b4257938c28a20d378e80e)) and different language RuntimeEnv classes(python [runtime_env.py](https://github.com/ray-project/ray/pull/25387/files#diff-df75b3b5ccad7d0b346d4a98396b66f305dc9c8301bfe72a3ed175dbeb6be671), java [RuntimeEnv.java](https://github.com/ray-project/ray/pull/25387/files#diff-689170794593769b55fa297f556a4cdbe813a64dfdfb42e485a8cab30844c232) and C++ [runtime_env.h](https://github.com/ray-project/ray/pull/25387/files#diff-7070cd92518c026c5fc71f24859f08119c68ef5aa93ecea3217fff81d29796a7)).
- Second, when a developer wants to add a environment support, such as pip, he/she should add:
    * [required] A plugin implementation [pip_plugin.py](https://github.com/ray-project/ray/pull/25387/files#diff-230b19d6c80a031df625b4a7111be08cad4e13f4e13d1325b3e568eec72f9a89)
    * [required]A json schema file which is used to validate the input data. [pip_schema.json](https://github.com/ray-project/ray/pull/25387/files#diff-22b4da3f0829b59660f1c518062645af04659f297371fb16bc16a6efbc352aa8)
    * [optional]A python [dataclass](https://docs.python.org/3/library/dataclasses.html) -> [pip.py](https://github.com/ray-project/ray/pull/25387/files#diff-5e636fca39c6a640b903659786561c7269f7ee3c0060a300f64609442b733511)
    * [optional]A java [POJO](https://en.wikipedia.org/wiki/Plain_old_Java_object) class -> [Pip.java](https://github.com/ray-project/ray/pull/25387/files#diff-5099898d0587b5b44c4f3a686836250eb70ec6179afe64ffba77efd0b92fc557)
    * [optional] A c++ class which could be converted by [nlohmann/json](https://github.com/nlohmann/json#basic-usage) -> [pip.h](https://github.com/ray-project/ray/pull/25387/files#diff-6c582782d7a62b3f864f5df0d4c83457b48fe17d5b213fd9da83da06fac8a6ec)
- Then, it should work.
    * Start a ray cluster with plugins
    ```
     ray start --head --runtime-env-plugins="{\"pip_plugin.py\": {\"schema\": \"pip_schema.json\", \"config\": {} }}"
    ```  
     * The behavior of runtime env agent
         - Load all the plugins using  `plugin_manager.py` and process setup runtime env by them.
     * The behavior of workers
         - When the workers are started, it should load all the json schema files(pip_schema.json) using plugin schema manager(plugin_schema_manager.py, PluginSchemaManager.java, plugin_schema_manager.h) and validate the user inputs by them.

     * The user code
         - When a user wants to use pip in python:
              * option 1: Use dict directly
                  ```
                      runtime_env = {"pip": {"packages": ["requests"]}}
                  ```
              * option 2: Get a released dataclass`pip.py` and use a entire strongly-typed API
                  ```
                      runtime_env = RuntimeEnv()
                      runtime_env.set("pip", Pip(packages=["requests"]))
                  ```

        - Primitive type in python:
             * option 1: Use dict directly
                  ```
                      runtime_env = {"working_dir": "https://path/to/dir.zip"}
                  ```
             * option 2: Use strongly-typed API
                  ```
                      runtime_env = RuntimeEnv()
                      runtime_env.set("working_dir", "https://path/to/dir.zip")
                  ```

         - When a user wants to use pip in Java:
              * option 1: Use json string  in the strongly-typed API
                  ```
                      RuntimeEnv runtimeEnv = new RuntimeEnv();
                      String pipString = "    "
                        + "{\n"
                        + "      \"packages\": [\"requests\"],\n"
                        + "}";
                      runtimeEnv.setJsonStr("pip", pipString);
                  ```
              * option 2: Get a released `Pip.java` and use a entire strongly-typed API
                  ```
                        RuntimeEnv runtimeEnv = new RuntimeEnv();
                        Pip pip = new Pip();
                        String[] packages = new String[] {"requests"};
                        pip.packages = packages;
                        runtimeEnv.set("pip", pip);
                  ```

        - Primitive type in java:
             * Use strongly-typed API
                  ```
                      RuntimeEnv runtimeEnv = new RuntimeEnv();
                      runtimeEnv.set("working_dir", "https://path/to/dir.zip");
                  ```

         - When a user wants to use pip in C++:
              * option 1: Use json string  in the strongly-typed API
                  ```
                      RuntimeEnv runtime_env;
                      std::string pip_raw_json_string = R"(
                        {
                          "packages": ["requests"],
                        }
                      )";
                      runtime_env.SetJsonStr("pip", pip_raw_json_string);
                  ```

              * option 2: Get a released `pip.h` and use a entire strongly-typed API
                  ```
                      RuntimeEnv runtime_env;
                      std::vector<std::string> packages = {"requests"};
                      Pip pip(packages, true);
                      runtime_env.Set("pip", pip);
                  ```

        - Primitive type in C++:
             * Use strongly-typed API
                  ```
                      RuntimeEnv runtime_env;
                      runtime_env.Set("working_dir", "https://path/to/dir.zip");
                  ```

Following this prototype, we should abandon the [protobuf](https://github.com/ray-project/ray/blob/master/src/ray/protobuf/runtime_env_common.proto) because it conflicts with pluggability.

________________________________________

## Progress tracing

This PR is only a prototype which won't be merged. The related formal PRs are tracing here:

- https://github.com/ray-project/ray/pull/26077
- https://github.com/ray-project/ray/pull/26154
- https://github.com/ray-project/ray/pull/26296
- https://github.com/ray-project/ray/pull/26522

### CODE REVIEW

**Comment by simon-mo** [comment]
> A python extension user API [pip.py](https://github.com/ray-project/ray/pull/25387/files#diff-5e636fca39c6a640b903659786561c7269f7ee3c0060a300f64609442b733511)
A java extension user API [Pip.java](https://github.com/ray-project/ray/pull/25387/files#diff-5099898d0587b5b44c4f3a686836250eb70ec6179afe64ffba77efd0b92fc557)
A c++ extension user API [pip.h](https://github.com/ray-project/ray/pull/25387/files#diff-6c582782d7a62b3f864f5df0d4c83457b48fe17d5b213fd9da83da06fac8a6ec)


Can this be easier? How about arbitrary JSON compatible type? That will be 
- `dict[str, Any]` in Python
-  `Map[String, String | List[String] | Map[String, ...]]` in Java or C++

**Comment by SongGuyang** [comment]
I think we need a strong typed API. Of course, we keep the compatibility of Dict in python. Users can use both strong type and dict/list/string... in python. But in java and c++, the strong type is more friendly.

**Comment by simon-mo** [comment]
Is there a way to lighten the load here? How does other plug-in author write these files?

**Comment by architkulkarni** [comment]
I think this API makes sense.  It's unfortunate that a plugin has to be bundled with these three manually written interface files (java, cpp, py) but I don't know of a better solution if we want to support multiple languages.  

I also have a very basic question, please bear with me-- naively it seems like protobuf should solve the problem of having to manually write an interface file for each language.  But I guess the problem there is that if the plugin just consists of a proto file, it would still have to be compiled into the different interface files, which can't be done at runtime.  And if on the other hand we ask the plugin author to autogenerate the interface files from the .proto and bundle those in the plugin, the result will be more complicated than just having 3 manually written interface files. Am I understanding that right?

**Comment by architkulkarni** [comment]
Actually, how does the pluggability work for C++ and Java?  For Python the class can just be loaded at runtime, but for C++ and java, when do the `.h` and `.java` files get compiled?

File: python/ray/_private/runtime_env_prototype/pluggability/plugin.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+            Returns:
+
+                disk_space(float): the disk space taken up by this plugin installation for this
+                environment. e.g. for working_dir, this downloads the files to the
+                local node.
+
+                Workerly(bool): A flag indicate that the runtime env is only used for one worker.
+
+                jobly(bool): A flag indicate that the runtime env is only used for one job.
```
[the remark this answers is not in the document]
Comment (SongGuyang) [reply python/ray/_private/runtime_env_prototype/pluggability/plugin.py:69]: > Would it make sense for these just be attributes of the Plugin? 

Make sense! I think we can make this two flag to be attributes.

> Also, can you remind me the use case of per_job? Is it a current or future use case? 

Yep, it is for future. Currently, I can propose one case: We can define a plugin named JobDataDirPlugin which will create a directory when job started, and store the job data for several days after job finished. The job can get the job directory and write data to it in runtime. It is a way to manage runtime data of job.

And I‘m also thinking about another case. If a job don't want to use the cached uris, how to implement? For example, a user submit a job with a remote working dir. After the job fails, he update the content of the same remote working dir and submit a new job.  The content are changed, but the uri is the same. In this case, the user can use a flag which indicate disable the cache for the job. But I think it is a flag for cache module, not plugins.

It's just brainstorming. We can add this flag if needed in future.

**Comment by SongGuyang** [comment]
> I also have a very basic question, please bear with me-- naively it seems like protobuf should solve the problem of having to manually write an interface file for each language. But I guess the problem there is that if the plugin just consists of a proto file, it would still have to be compiled into the different interface files, which can't be done at runtime. And if on the other hand we ask the plugin author to autogenerate the interface files from the .proto and bundle those in the plugin, the result will be more complicated than just having 3 manually written interface files. Am I understanding that right?

I think your understanding is right! Actually, protobuf is also an alternative. The benefit of protobuf is that it could make the protocol between plugin and user API clear. That's the reason why we use it before. But If some plugins' interface is simple, e.g. working_dir, using protobuf will make it complicated. Maybe it is a trade-off? 

If we can find a easy way to use protobuf, I think we can use it.

**Comment by SongGuyang** [comment]
@architkulkarni @simon-mo About the way to use a plugin, I can explain anymore.

We assume the ray version 2.0 has been released without the pip support. 
After that, person A develops the pip_plugin.py and releases it. 

Person B wants to use pip in python, so he use it like:
- start cluster:
```
ray start --head --runtime-env-plugins "pip_plugin.py"
```
- driver code:
```
runtime_env = {
    "pip": {
         "packages": ["requests"],
          "pip_check": True
    }
}
```

Person C also wants to use pip in python, but he want to use strong-typed API. So he develops a module named pip.py and use it:
- start cluster:
```
ray start --head --runtime-env-plugins "pip_plugin.py"
```
- driver code:
```
runtime_env = RuntimeEnv()
runtime_env.set("pip", Pip(packeges=["requests"], pip_check=True)
```

Person D wants to use pip in Java for xlang scenes. So he develops a module named Pip.java and use it:
- start cluster:
```
ray start --head --runtime-env-plugins "pip_plugin.py"
```
- driver code:
```
RuntimeEnv runtimeEnv = new RuntimeEnv();
Pip pip = new Pip();
String[] packages = new String[] {"requests"};
pip.packages = packages;
pip.pip_check = true;
runtimeEnv.set("pip", pip);
```

Person E wants to use pip in C++ for xlang scenes. So he develops a module named pip.h and use it:
- start cluster:
```
ray start --head --runtime-env-plugins "pip_plugin.py"
```
- driver code:
```
RuntimeEnv runtime_env;
std::vector<std::string> packages = {"requests"};
Pip pip(packages, true);
runtime_env.Set("pip", pip);
```

And Then, in ray version 2.1, person A, C, D, E contribute their code(pip_plugin.py, pip.py, Pip.java, pip.h) to the ray project. After version 2.1 released, all the users could use pip in any language.

Overall, If you want to use the pip plugin, pip_plugin.py is necessary, but  pip.py, Pip.java and pip.h are optional.  pip.py, Pip.java and pip.h could be in the ray project, or in the user' projects.  So, for C++ and Java, the sdk class(Pip.java and pip.h) should be complied in the user's projects, not in the ray project.

**Comment by SongGuyang** [comment]
In addition, the sdk classes are simple. They are just data classes. It's so easy to add.

**Comment by architkulkarni** [comment]
Thanks for the clear explanation and for clarifying that the SDK classes are optional! So for the basic functionality (Python-only, no strong-type) the implementation load is the same as before, which is to just provide the single file `pip_plugin.py` like you said.  I think this addresses the concern about the implementation load being too high.

**Comment by SongGuyang** [comment]
We also need to enhance the [plugin_manager.py](https://github.com/ray-project/ray/pull/25387/files#diff-5862b6ec93c00728bf65cb7761c2dee9496d76717a0eda30889e18ef9d5b67cf). I think we can implement configuration in it. For example, pass some configs to each plugins. Or, control the conflict rules of plugins, e.g. pip and conda couldn't be used at the same time.

**Comment by simon-mo** [comment]
How will the `Pip` class get distributed if they don't live in the main Ray repo? (afterall, these are supposed for plugins?) Does this mean plugin authors need to distribute all the cross language headers, jars, and packages as well?

**Comment by simon-mo** [comment]
I would strongly encourage to find a way to reduce this kind of authoring load and usage overhead. Two alternatives:
- As you mentioned protobuf can be useful here. 
- How about JSON schema definition? It has robust library in all languages and support validation

**Comment by SongGuyang** [comment]
> How will the `Pip` class get distributed if they don't live in the main Ray repo? (afterall, these are supposed for plugins?) Does this mean plugin authors need to distribute all the cross language headers, jars, and packages as well?

Plugin authors could distribute all the cross language headers, jars, and packages if they want. But if they only want to distribute the `pip_plugin.py`, other users can add the  pip.py, Pip.java and pip.h to their local project by themself.

**Comment by SongGuyang** [comment]
> I would strongly encourage to find a way to reduce this kind of authoring load and usage overhead. Two alternatives:
> 
> * As you mentioned protobuf can be useful here.
> * How about JSON schema definition? It has robust library in all languages and support validation

Can you show some details? Or can you add some code to this prototype directly? We can collaborate to make this prototype.

**Comment by SongGuyang** [comment]
I will try to learn JSON schema definition next week. Thanks for this advise.

**Comment by SongGuyang** [comment]
@simon-mo Do you mean that we can use a json schema library to generate the codes of different languages from a json schema format? I found a library which looks like it can do it https://app.quicktype.io/#l=python. But I'm not sure if we can use it without usage overhead.

**Comment by simon-mo** [comment]
Can you elaborate on the overhead? JSON Schema is basically an interface definition language that's language dependent. 

Here's the three levels support I imagine here, ranked from complexity low->high:

### Level 1: Perform validation in agent
Just do the validation in agent, no need to typed API. Plugin author only needs to write the validation logic in their plugin class. This means one RPC on each new runtime env (you can skip validation for repeated run via hashing right?)

### Level 2: Plugin publishes shema.json, validate in each language, code gen optional.
Plugin author can write a JSON schema description of the accepted type. for example the author can add the following schema for `pip`:

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "pip": {
      "type": "object",
      "properties": {
        "packages": {
          "type": "array",
          "items": [
            {
              "type": "string"
            },
            {
              "type": "string"
            }
          ]
        },
        "pip_check": {
          "type": "boolean"
        }
      },
      "required": [
        "packages",
        "pip_check"
      ]
    }
  },
  "required": [
    "pip"
  ]
}
```

and each language can validate them with the schema. JSON schema validators are well supported in each language and they are production level: https://json-schema.org/implementations.html These library does runtime validation really well.

For code generation I think they should be strictly optional, and they shouldn't be part of the plugin author critical path.

### Level 3: Author writes Pydantic classes as source of truth
The plugin author should be recommended to use [Pydantic](https://pydantic-docs.helpmanual.io/) to automatically generate the right JSON schema.

**Comment by SongGuyang** [comment]
@simon-mo In your summary, my current prototype is level 1? There is a validation interface in plugin https://github.com/ray-project/ray/pull/25387/files#diff-230b19d6c80a031df625b4a7111be08cad4e13f4e13d1325b3e568eec72f9a89R19.

**Comment by SongGuyang** [comment]
> For code generation I think they should be strictly optional, and they shouldn't be part of the plugin author critical path.

@simon-mo about this, if we don't generate language code, how can we achieve the API like(use Java for example):
```
      RuntimeEnv runtimeEnv = new RuntimeEnv();
      Pip pip = new Pip();
      String[] packages = new String[] {"requests"};
      pip.packages = packages;
      pip.pip_check = true;
      runtimeEnv.set("pip", pip);
```

**Comment by simon-mo** [comment]
In case of Java, you can have users defining their inline classes in Java; or just JSON schema to generate the class type. As long as they can be serialized to JSON.  

My point is that we shouldn't force language's typing constraint here. For example in C++, you don't need a strongly typed API, users can just nlohmann::json's json class easily without typing to create this kind of nested structure.

**Comment by SongGuyang** [comment]
@simon-mo I see. Let me rethink. I have another question. Do you think we should keep a big json? It mean that the user input should be a json object(e.g. nlohmann::json), not a json string. The difference of the final json:

use json object(big json):
```
{
    "pip" : {
        "packages": ["requests"]
    }
}
```
use json string:
```
{
    "pip" : "{\"packages\": [\"requests\"]}"
}
```

I keep the big json in current prototype because:
- Keep the compatibility with the dict API in python https://github.com/ray-project/ray/pull/25387/files#diff-a37369be55591323e18e53fb7d15abb3f970fe66b75998835ef410ac929f7d93R26
- Easy to read if we print the serialized runtime env

But the cons of big json is that the json library will be exposed to users https://github.com/ray-project/ray/pull/25387/files#diff-6c582782d7a62b3f864f5df0d4c83457b48fe17d5b213fd9da83da06fac8a6ecR18. And it will make our implementation complex.

Wdyt?

**Comment by SongGuyang** [comment]
@simon-mo @architkulkarni I have added the json schema validation in this prototype. In worker side, we can use `plugin_schema_manager.py`(plugin_schema_manager.h in C++) load all the plugin schemas and validate the user inputs.

**Comment by SongGuyang** [comment]
@simon-mo @architkulkarni I have supported setting runtime env with raw json and raw json string in [C++](https://github.com/ray-project/ray/pull/25387/commits/b0907e22a2c0562bfbe0e91ac1a7cef7b3637c59) and [Java](https://github.com/ray-project/ray/pull/25387/commits/94ca807062f8ce22f543430dc816817afa40ff45). It means that the plugin authors don't need to private the `pip.py/pip.h/Pip.java` classes in different languages. `pip.py/pip.h/Pip.java` could be optional. 

@simon-mo Does this prototype is already like what you said?

**Comment by SongGuyang** [comment]
@edoakes @simon-mo @architkulkarni The description has been updated. I think it will make the prototype more clear.

**Comment by SongGuyang** [comment]
@edoakes @simon-mo @architkulkarni The description has been updated. I think it will make the prototype more clear.

File: python/setup.py [hunk 2]
```diff
@@ -289,6 +289,8 @@ def get_packages(self):
         "frozenlist",
         "requests",
         "virtualenv",  # For pip runtime env.
+        # We need to discuss if there is other alternative for converting dict to dataclass.
+        "dacite",
```
Comment (SongGuyang) [inline python/setup.py:293]: I import `dacite` to achieve converting `dict` to `dataclass`. We can also find a better way to do this if possible.

**Comment by SongGuyang** [comment]
@edoakes @architkulkarni @simon-mo No comments?

**Comment by architkulkarni** [comment]
Very sorry for the delay, the next step is for me to review this. Let me
review this today.

**Comment by architkulkarni** [comment]
Thanks for the detailed update @SongGuyang! Let me try to summarize, and please let me know if anything sounds wrong:
1. In the PR description we pass in the plugins and schemas in `ray start`.  Is this a fundamental restriction?  A stronger form of pluggability is that the user can specify a new plugin in `ray.init(“ray://..”, runtime_env=)` or `ray job submit —runtime-env-json={…}`.  I think this should actually be a requirement for pluggability.  I think it would be useful for the use case of a long-running cluster with multiple users.  What do you think?
2. Summarizing the advantage of the new approach vs the previous approach in this PR: Now, users for all 3 languages don't have to use `.py` `.java` `.cc` interface files, as long as they are OK with passing in JSON.  The interface files are only required for strong type.  This makes sense to me and I believe it addresses Simon's concern.  The new approach is Level 2 in Simon's hierarchy.
3. Regarding `dacite`, it's small and has an MIT license so I think we should copy it into the Ray repo to reduce the number of Ray dependencies, similar to what we've done for `pathspec`.
4. Regarding [big JSON vs JSON string](https://github.com/ray-project/ray/pull/25387#issuecomment-1149400147) -- sorry if this is outdated due to my delayed review. The current prototype uses jig JSON and one of the cons listed was that it makes the implementation more complex. I saw you linked to `to_json` and `from_json` in `pip.h` as an example, but I didn’t see those functions called anywhere.  Is this superseded by the to/from json functions in `RuntimeEnv.java`, `runtime_env.h`, etc?  If that's the case I don't think the added complexity is too bad.

Overall, this looks good to me and I'd be comfortable moving forward with this design, provided we can figure out a solution for (1).

**Comment by SongGuyang** [comment]
@architkulkarni Thanks for your comments. Let me reply your comments first.
1. I think static loading plugins when ray cluster starts could cover most of the scenes. But I'd like to discuss how to support dynamic loading if you think it is useful and feasible. Some issues I can propose:
    - a. How can we find the plugin classes by plugin name in runtime? Do you mean the users input a full path of plugin in the code like `ray.init(“ray://..”, runtime_env={"/path/to/pip_plugin.py": ...})`? It will make the schema inconsistent.
    - b. How can we make configs for the dynamic loaded plugins? For example, we should config the conflict strategy and orders for all of the plugins. In my thought, we should make configs when the ray cluster starts. I haven't design this part in current prototype yet, but we need this.
 
    In my current design, the plugins are managed in the cluster level. If all the plugins could be loaded dynamically, it will be different entirely.

2. Right!

3. Yep, I can try to copy it in the following PRs, let's discuss it later in the concrete PR.

4. I think we can ignore this issue now. The complexity is acceptable for us.

Overall, I think we can discuss for (1) continuously. But we can start to make some PRs at the same time. The steps in my thought are that:
- step 1: Remove the protobuf.
- step 2: Add the plugin framework and make the existing plugin tests work.
- step 3: Covert the existing runtime env modules to plugins, e.g. pip, conda, container, working_dir, java_jars...
- step 4: Address plugins dynamic loading issue if we already find a way to do this.

Wdyt?

**Comment by architkulkarni** [comment]
For (1), the original vision for pluggability included (for example) the ability to allow a plugin to replace `working_dir` plugin by a faster plugin that used Amazon S3 for the storage instead of the GCS, or to use PEX to load a python package.  There the API was simply `runtime_env: {"my_pkg.my_plugin_class": my_plugin_config}`, though at that time we weren't considering these cross-language issues.  Now that the validation is part of the schema file instead of a plugin class method, I guess it'll have to be passed in some other way.

Maybe we should stipulate that if you want to use a dynamically loaded plugin together with cross-language, you must use pass in JSON instead of strong typing?  I didn't see an obvious way to make C++ strong typing work with a dynamically specified plugin, but I might be missing something.

If you're pretty sure the proposed PRs won't prevent us from supporting (1) by painting us into a corner, I'm happy to move forward with those PRs while continuing to discuss (1).  It's hard for me to predict though, it seems a bit complicated.

(One other thought: it would be a great user experience if we could ship `my_plugin_class` from the local machine to all the nodes on the Ray cluster, instead of requiring it to be present on all nodes beforehand. Perhaps it could be done somehow using `working_dir`, but that doesn't seem possible with the current order of operations... I'll have to think more about this)

**Comment by SongGuyang** [comment]
@architkulkarni For your requirements, maybe we can support one way:

- Load the pre-defined plugins when ray cluster starts:
ray start --head --runtime-env-plugins="{\"pip_plugin.py\": {\"schema\": \"pip_schema.json\", \"config\": {} }}"

- Add new plugin dynamically for the running cluster:
ray update --runtime-env-plugins="{\"/local/path/to/pex_plugin.py\": {\"schema\": \"/local/path/to/pex_schema.json\", \"config\": {} }}"

Some thing like this. The `ray update` command will ship the local files(pex_plugin.py and pex_schema.json) to cluster nodes and talk agents to load the new plugin. 

We should add a module in GCS to manage the plugins and schemas because we need to ship the files to new nodes which created by autoscaling or failover.

> Maybe we should stipulate that if you want to use a dynamically loaded plugin together with cross-language, you must use > pass in JSON instead of strong typing? I didn't see an obvious way to make C++ strong typing work with a dynamically  specified plugin, but I might be missing something.

I think there are no issues for cross-language because the strong-type class are compiled in user(job) codes which the ray
cluster is not aware of.

**Comment by architkulkarni** [comment]
OK, that sounds good.  After thinking more I think you're right that defining plugins at cluster startup covers most situations, so we can add the dynamic case later if necessary.

**Pull request closed without merging**